### PR TITLE
[New Feature] Multilingual: Propagating existing associations if desired

### PR DESCRIPTION
**Pull request by infograf768**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/21250

### Summary of Changes
This adds a "Propagate" button in the Associations tab when editing a contact, category, article, menu item, newsfeed.
This button will display when an association is selected in a Content Language field, offering the option to Propagate the associations defined for that item to the other Content Languages.

The button will display ONLY when there are more than 2 Content Languages defined for the site.
A message will display with the results obtained.

It is B/C as it will not break anything and one is totally free to choose or not to use the button, thus keeping the possibility to Clear/Select/Create any associations as before.

Using it prevents breaking existing associations when the wrong items are chosen for some fields.

A big thank you to **Robbie Jackson** for the long work to get this new functionality.
Target Milestone: 3.9.0

### Testing Instructions
Create a Multilingual site with 3 or more languages.
Create items in each type: (contact, category, article, menu item, newsfeed) for 2 of these languages and associate them in each case.

Create a new item for each type, select another language than the ones which are already associated.
Save (NOT save and close) this item.
Display the Association tab.

You will see, as usual (here for 4 Content Languages)

screen shot 2018-07-31 at 11 13 04

I do have already associations set for English and French.

I select for English the item which I know is associated.

The button displays (Better tip welcome)

screen shot 2018-07-31 at 11 17 17

Clicking on the button will give

screen shot 2018-07-31 at 11 18 22

Result: the already associated category in French is added in the French field.
As we have no association set for Spanish, the field remains empty.
The message displays (it is the same in all cases except if there is an Ajax error)
`All existing associations have been set. If some are missing you may have to select or create them manually.`

If we have already associated all fields, clicking on Propagate will just display the message.

If we change an association in one of the fields and click Propagate, it is the associations from that language item that will be propagated and therefore replace the ones existing.

### Documentation Changes Required
Yes

### CODE REVIEW

**Comment by infograf768** [comment]
the json files need corrections for cs. Can be tested and commented nevertheless. ;)

File: administrator/language/en-GB/en-GB.ini [hunk 1]
```diff
@@ -281,6 +281,10 @@ JGLOBAL_ARTICLE_ORDER_DESC="The order that articles will show in."
 JGLOBAL_ARTICLE_ORDER_LABEL="Article Order"
 JGLOBAL_ARTICLES="Articles"
 JGLOBAL_ASSOC_NOT_POSSIBLE="To define associations, please make sure the item language is not set to 'All'."
+JGLOBAL_ASSOCIATIONS_PROPAGATE_BUTTON="Propagate"
+JGLOBAL_ASSOCIATIONS_PROPAGATE_FAILED="Failed propagating associations. You may have to select or create them manually."
+JGLOBAL_ASSOCIATIONS_PROPAGATE_MESSAGE="All existing associations have been set. If some are missing you may have to select or create them manually."
+JGLOBAL_ASSOCIATIONS_PROPAGATE_TIP="Propagates this selected item existing associations."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini]: Please correct to

"Propagates this item's existing associations."

File: administrator/language/en-GB/en-GB.ini [hunk 2]
```diff
@@ -281,6 +281,10 @@ JGLOBAL_ARTICLE_ORDER_DESC="The order that articles will show in."
 JGLOBAL_ARTICLE_ORDER_LABEL="Article Order"
 JGLOBAL_ARTICLES="Articles"
 JGLOBAL_ASSOC_NOT_POSSIBLE="To define associations, please make sure the item language is not set to 'All'."
+JGLOBAL_ASSOCIATIONS_PROPAGATE_BUTTON="Propagate"
+JGLOBAL_ASSOCIATIONS_PROPAGATE_FAILED="Failed propagating associations. You may have to select or create them manually."
+JGLOBAL_ASSOCIATIONS_PROPAGATE_MESSAGE="All existing associations have been set. If some are missing you may have to select or create them manually."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini]: How likely is it that an association will be missing? Seems odd to me to have that on a success message.

File: administrator/language/en-GB/en-GB.ini [hunk 3]
```diff
@@ -281,6 +281,10 @@ JGLOBAL_ARTICLE_ORDER_DESC="The order that articles will show in."
 JGLOBAL_ARTICLE_ORDER_LABEL="Article Order"
 JGLOBAL_ARTICLES="Articles"
 JGLOBAL_ASSOC_NOT_POSSIBLE="To define associations, please make sure the item language is not set to 'All'."
+JGLOBAL_ASSOCIATIONS_PROPAGATE_BUTTON="Propagate"
+JGLOBAL_ASSOCIATIONS_PROPAGATE_FAILED="Failed propagating associations. You may have to select or create them manually."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini:285]: Shouldn't this be "will" not "may"?

File: administrator/language/en-GB/en-GB.ini (hunk 2 above)
Comment (infograf768) [reply administrator/language/en-GB/en-GB.ini]: It will happen when there is no item associated in a specific language. See image above where there is nothing for Spanish

Comment (brianteeman) [reply administrator/language/en-GB/en-GB.ini]: so in that case nothing is missing as it doesnt exist
The first sentence is more than enough

File: administrator/language/en-GB/en-GB.ini (hunk 3 above)
Comment (infograf768) [reply administrator/language/en-GB/en-GB.ini:285]: Some one may not want to create associations for a specific language.
Change to ?
`If desired, you will have to select or create them manually.`

File: administrator/language/en-GB/en-GB.ini (hunk 2 above)
Comment (infograf768) [reply administrator/language/en-GB/en-GB.ini]: This message is also informing the user that an association for one or more language(s) may be missing. It may be what the user wants but it also may not be.
In that case the user can select an item or create one directly in this interface.
This is important.

File: administrator/language/en-GB/en-GB.ini (hunk 3 above)
Comment (brianteeman) [reply administrator/language/en-GB/en-GB.ini:285]: then they havent failed and this isnt a failure message

File: administrator/language/en-GB/en-GB.ini (hunk 2 above)
Comment (brianteeman) [reply administrator/language/en-GB/en-GB.ini]: thats not the role of a success message.

Comment (infograf768) [reply administrator/language/en-GB/en-GB.ini]: It is only a success message for what could be propagated.

File: administrator/language/en-GB/en-GB.ini (hunk 3 above)
Comment (infograf768) [reply administrator/language/en-GB/en-GB.ini:285]: Propagate has failed (for an unknown reason. It may be the case for a 3rd party component which is badly coded). It does not mean the user lost the possibility to associate items by selecting/creating them.

I am now waiting for tests. Thank you.

**Comment by brianteeman** [comment]
Testing this and my only issue is with the messages

The propagate button says
> Propagate this items existing associations

But you always get this message
> All existing associations have been set. If some are missing you may have to select or create them manually.

The code should check to see if there any associations to propagate and if not then the message should be
> No associations exist to propagate

And if there are associations to propagate then the message should be
> All existing associations have been set.

**Comment by infograf768** [comment]
We will now get 3 types of messages:
JGLOBAL_ASSOCIATIONS_PROPAGATE_MESSAGE_ALL="All existing associations have been set."
JGLOBAL_ASSOCIATIONS_PROPAGATE_MESSAGE_NONE="No associations exist to propagate."
JGLOBAL_ASSOCIATIONS_PROPAGATE_MESSAGE_SOME="Some associations have been set: %s"

JGLOBAL_ASSOCIATIONS_PROPAGATE_MESSAGE_SOME will also display the language(s) where associations have been added.

Example. Here I selected an article for English. It added the French existing association, but as the Spanish one did not exist, it is that message which is used displaying which association was set.
screen shot 2018-07-31 at 16 34 41

**Comment by infograf768** [comment]
Restarted drone who failed on js.

File: administrator/language/en-GB/en-GB.ini [hunk 4]
```diff
@@ -281,6 +281,11 @@ JGLOBAL_ARTICLE_ORDER_DESC="The order that articles will show in."
...
+JGLOBAL_ASSOCIATIONS_PROPAGATE_MESSAGE_ALL="All existing associations have been set."
+JGLOBAL_ASSOCIATIONS_PROPAGATE_MESSAGE_NONE="No associations exist to propagate."
+JGLOBAL_ASSOCIATIONS_PROPAGATE_MESSAGE_SOME="Some associations have been set: %s"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.ini]: Please change to 
Associations have been set for: %s

**Comment by brianteeman** [comment]
Much better now. 

I am also getting a missing string JGLOBAL_ASSOCIATIONS_PROPAGATE_TIP

**Comment by infograf768** [comment]
done

File: administrator/language/en-GB/en-GB.ini (hunk 3 above)
Comment (brianteeman) [reply administrator/language/en-GB/en-GB.ini:285]: ok change to your suggestion of "If desired, you will have to select or create them manually.

**Comment by jreys** [comment]
I have tested this item :white_check_mark: successfully on eabc256e01ab0b784a8e101b6a2e8fb05b5e2eb3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21321">issues.joomla.org/tracker/joomla-cms/21321</a>.</sub>

**Comment by infograf768** [comment]
One more tester.

File: administrator/components/com_categories/controllers/ajax.json.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+			JTable::addIncludePath(JPATH_ADMINISTRATOR . '/components/com_categories/tables');
+			$categoryTable = JTable::getInstance('Category', 'JTable');
+
+			foreach ($associations as $lang => $association)
+			{
+				$categoryTable->load($association->id);
+				$associations[$lang]->title = $categoryTable->title;
+			}
+
+			$message = null;
```
Comment (Quy) [inline administrator/components/com_categories/controllers/ajax.json.php]: Remove. Not necessary as it will be assigned later in the if statement.

Comment (infograf768) [reply administrator/components/com_categories/controllers/ajax.json.php]: Waiting for other comments from you before changing. 👍

Comment (Quy) [reply administrator/components/com_categories/controllers/ajax.json.php]: Repeat this change in the other files.

File: administrator/components/com_newsfeeds/models/fields/modal/newsfeed.php [hunk 6]
```diff
@@ -169,6 +176,23 @@ function jSelectNewsfeed_" . $this->id . "(id, title, object) {
 				. '</a>';
 		}
 
+		// Propagate contact button
```
Comment (Quy) [inline administrator/components/com_newsfeeds/models/fields/modal/newsfeed.php]: Update comment.

**Comment by infograf768** [comment]
@Quy 
All done. Thanks.

**Comment by infograf768** [comment]
@Quy @franz-wohlkoenig @brianteeman 
Please add your tests.

**Comment by ghost** [comment]
@infograf768 have read Instructions a few Times but didn't understand really what is to test. So i need Time.

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on b902d37eacec211d9b2621241bdf7b8f842d1296<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21321">issues.joomla.org/tracker/joomla-cms/21321</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
Ready to Commit after two successful tests.

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by infograf768** [comment]
@mbabker 
I have tagged this to 3.9. Can you merge?

File: administrator/components/com_categories/views/category/tmpl/edit.php [hunk 7]
```diff
@@ -101,6 +101,7 @@
 		<?php echo $this->form->getInput('extension'); ?>
 		<input type="hidden" name="task" value="" />
 		<input type="hidden" name="forcedLanguage" value="<?php echo $input->get('forcedLanguage', '', 'cmd'); ?>" />
+		<?php echo '<input id="token" type="hidden" name="' . JSession::getFormToken() . '" value="1" />'; ?>
```
Comment (mbabker) [inline administrator/components/com_categories/views/category/tmpl/edit.php]: This line shouldn't be necessary, this is exactly what the JHtml call immediately below does.

File: administrator/components/com_contact/views/contact/tmpl/edit.php [hunk 8]
```diff
@@ -120,5 +120,6 @@
 	</div>
 	<input type="hidden" name="task" value="" />
 	<input type="hidden" name="forcedLanguage" value="<?php echo $input->get('forcedLanguage', '', 'cmd'); ?>" />
+	<?php echo '<input id="token" type="hidden" name="' . JSession::getFormToken() . '" value="1" />'; ?>
```
Comment (mbabker) [inline administrator/components/com_contact/views/contact/tmpl/edit.php]: Same as above.

File: administrator/components/com_content/views/article/tmpl/edit.php [hunk 9]
```diff
@@ -138,6 +138,7 @@
 		<input type="hidden" name="task" value="" />
 		<input type="hidden" name="return" value="<?php echo $input->getCmd('return'); ?>" />
 		<input type="hidden" name="forcedLanguage" value="<?php echo $input->get('forcedLanguage', '', 'cmd'); ?>" />
+		<?php echo '<input id="token" type="hidden" name="' . JSession::getFormToken() . '" value="1" />'; ?>
```
Comment (mbabker) [inline administrator/components/com_content/views/article/tmpl/edit.php]: Same as above.

File: administrator/components/com_menus/views/item/tmpl/edit.php [hunk 10]
```diff
@@ -232,6 +232,7 @@
 
 	<input type="hidden" name="task" value="" />
 	<input type="hidden" name="forcedLanguage" value="<?php echo $input->get('forcedLanguage', '', 'cmd'); ?>" />
+	<?php echo '<input id="token" type="hidden" name="' . JSession::getFormToken() . '" value="1" />'; ?>
```
Comment (mbabker) [inline administrator/components/com_menus/views/item/tmpl/edit.php]: Same as above.

File: administrator/components/com_newsfeeds/views/newsfeed/tmpl/edit.php [hunk 11]
```diff
@@ -107,5 +107,6 @@
 	</div>
 	<input type="hidden" name="task" value="" />
 	<input type="hidden" name="forcedLanguage" value="<?php echo $input->get('forcedLanguage', '', 'cmd'); ?>" />
+	<?php echo '<input id="token" type="hidden" name="' . JSession::getFormToken() . '" value="1" />'; ?>
```
Comment (mbabker) [inline administrator/components/com_newsfeeds/views/newsfeed/tmpl/edit.php]: Same as above.

File: media/system/js/associations-edit-uncompressed.js [hunk 12]
```diff
@@ -30,6 +30,104 @@ window.showAssociationMessage = function()
...
+	window.propagateAssociation = function(fieldPrefix, callbackFunctionPrefix) 
+	{
+		// Find the id of the record which has been set as an assocation
+		var assocId = jQuery("#" + fieldPrefix + "_id").val();
+
+		// Find the language of the record being edited
+		var currentLang = jQuery('#jform_language').find(":selected").val();
+
+		// Find the token so that it can be sent in the Ajax request as well
+		var token = jQuery("#token").attr("name");
```
Comment (mbabker) [inline media/system/js/associations-edit-uncompressed.js]: Instead of finding the token in this way, the API added in https://github.com/joomla/joomla-cms/pull/14952 should be used instead.

Comment (infograf768) [reply media/system/js/associations-edit-uncompressed.js]: You mean using
`var token = jQuery('meta[name=csrf-token]').val();` instead?

Comment (infograf768) [reply media/system/js/associations-edit-uncompressed.js]: hmm, that does not work

Comment (mbabker) [reply media/system/js/associations-edit-uncompressed.js]: In the layouts, you have to add a `JHtml::_('jquery.token');` call, then you don't even need to find the token and manually include it when using jQuery.ajax because that PHP call will set up jQuery to include the token automatically.

Comment (infograf768) [reply media/system/js/associations-edit-uncompressed.js]: This works.
var token = Joomla.getOptions('csrf.token', '');

Is that OK, @mbabker ?

Comment (mbabker) [reply media/system/js/associations-edit-uncompressed.js]: Yes.

Comment (infograf768) [reply media/system/js/associations-edit-uncompressed.js]: I did not add the `<?php echo JHtml::_('jquery.token'); ?>`
Just deleted the extraneous lines in the edit.php and used 
var token = Joomla.getOptions('csrf.token', '');

As I did not know what would be the consequences 

Can we go with that?

**Comment by infograf768** [comment]
@mbabker 
Do we need new tests?

File: media/system/js/associations-edit-uncompressed.js [hunk 13]
```diff
@@ -30,6 +30,104 @@ window.showAssociationMessage = function()
...
+	 *                                              with the data field containing the associations
+	 * @param   string      fieldPrefix             The stem of the html ids for the elements comprising the modal field
+	 * @param   string      callbackFunctionPrefix  The name of the callback function which the modal window uses to set the 
+	 *                                              selected item as the association, but minus the language tag at the end
+	 *
+	 * @return  boolean
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	window.injectAssociations = function(result, callbackFunctionPrefix)
```
Comment (dgrammatiko) [inline media/system/js/associations-edit-uncompressed.js]: This is really bad. Please name this as `Joomla.injectAssociations`. Global variables, functions etc are as bad in Javascipt as they are in PHP 👎

**Review by dgrammatiko — CHANGES_REQUESTED** [review]

File: media/system/js/associations-edit-uncompressed.js (hunk 13 above)
Comment (infograf768) [reply media/system/js/associations-edit-uncompressed.js]: I guess Robbie used this as we have above existing
window.hideAssociation
and 
window.showAssociationMessage

Comment (infograf768) [reply media/system/js/associations-edit-uncompressed.js]: I guess I found what has to be modifed to fit as both the functions use window.
It is not the purpose of this patch to also correct the existing ones.

Comment (infograf768) [reply media/system/js/associations-edit-uncompressed.js]: @dgrammatiko 
Do you agree with my statement?
If yes, I will do the patch right now

**Comment by infograf768** [comment]
@dgrammatiko 
Modified the new js functions name as required.
We can't obviously normalize the others as it would not be B/C.


@franz-wohlkoenig @jreys 
Please test again. :)

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 888a2807aeb4373cbab7c1464e035df3df7ce8dd<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21321">issues.joomla.org/tracker/joomla-cms/21321</a>.</sub>

**Comment by infograf768** [comment]
@mbabker Can you now merge?

**Comment by infograf768** [comment]
@mbabker 
As it is a new feature, shall not it be merged in 3.9.0 ?

**Comment by mbabker** [comment]
💩

**Comment by mbabker** [comment]
Branch merged to 3.9-dev via https://github.com/joomla/joomla-cms/commit/11153f720bad7886e990c3b3af5bbf4ca4cf4c00

**Comment by infograf768** [comment]
Thanks.
